spec/index.html

```diff
@@ -10344,25 +10342,46 @@ <h3>Evaluation Semantics</h3>
           </div>
           <div class="defn">
             <p><b>Definition: <span id="defn_evalGraph">Evaluation of Graph</span></b></p>
-            <pre class="code nohighlight">
-if <var>IRI</var> is a graph name in <var>D</var>
-    <a href="#defn_eval" class="evalFct">eval</a>( <var>D</var>(<var>G</var>), Graph(<var>IRI</var>,<var>P</var>) ) = <a href="#defn_eval" class="evalFct">eval</a>( <var>D</var>(<var>D</var>[<var>IRI</var>]), <var>P</var> )
-            </pre>
-            <pre class="code nohighlight">
-              if <var>IRI</var> is not a graph name in <var>D</var>
-                  <a href="#defn_eval" class="evalFct">eval</a>( <var>D</var>(<var>G</var>), Graph(<var>IRI</var>,<var>P</var>) ) = the empty multiset
-            </pre>
-            <pre class="code nohighlight">
-<a href="#defn_eval" class="evalFct">eval</a>( <var>D</var>(<var>G</var>), Graph(<var>var</var>,<var>P</var>) ) =
-    Let <var>R</var> be the empty multiset
-    foreach IRI <var>i</var> in <var>D</var>
-        <var>R</var> := <a href="#defn_algUnion" class="algFct">Union</a>(<var>R</var>, <a href="#defn_algJoin" class="algFct">Join</a>( <a href="#defn_eval" class="evalFct">eval</a>(<var>D</var>(<var>D</var>[<var>i</var>]), <var>P</var>) , Ω(<var>var</var>-&gt;<var>i</var>) ) )
-    the result is <var>R</var>
-            </pre>
+            <p>For every |x| that is
+              an <a data-cite="RDF12-CONCEPTS#dfn-IRI">IRI</a> or
```

Comment:
```suggestion
              an <a data-cite="RDF12-CONCEPTS#dfn-IRI">IRI</a>,
              a <a data-cite="RDF12-CONCEPTS#dfn-blank-node">blank node</a> or
```

Should the definition of the evaluation of "graph" mention blank nodes?

It isn't necessary because they come in via variables but:

1. It more clearly indicates that literal and triple terms can not be used
2. The intuitive of variable substitution applies

Reply:
@afs your edit suggestion is incorrect. The _x_ in Graph(_x_, _P_) may indeed be only an IRI or a variable; not a blank node. While the algebraic syntax is not explicitly defined in the spec, it is clear from the grammar of the user-facing syntax that there can be no blank node---see the [`GraphGraphPattern`production](https://www.w3.org/TR/sparql12-query/#rGraphGraphPattern):
```
[62] GraphGraphPattern ::= 'GRAPH' VarOrIri GroupGraphPattern
```
> Should the definition of the evaluation of "graph" mention blank nodes?
>
> It isn't necessary because they come in via variables but:
>
> 1. It more clearly indicates that literal and triple terms can not be used

Cannot be used for what exactly?

If you mean that they cannot be used as the _x_ in the new version of this definition, I think stating that _x_ may be an IRI or a variable (as done in this PR) is sufficient.

If you mean that they cannot be used as bindings for the variable _x_ (in case _x_ is a variable), I don't think that saying something like this belongs in this definition; maybe better suited for the informal text in Section [13.3 Querying the Dataset](https://www.w3.org/TR/sparql12-query/#queryDataset). Also, it is not entirely correct to say that literals and term maps cannot be assigned to variable _x_ in incoming solution mappings; such solution mappings would just not survive the Join within the foreach loop. Perhaps one thing to make this more obvious would be to extend the foreach-line of the algorithm from (current form)

foreach [graph name](https://www.w3.org/TR/rdf12-concepts/#dfn-graph-name) _gn_ in _D_

to

foreach [graph name](https://www.w3.org/TR/rdf12-concepts/#dfn-graph-name) _gn_ in _D_ (recall that a graph name may be an IRI or a blank node)

Would that address your concern?

> 2. The intuition of variable substitution applies

I am not sure what you mean by this comment.

Reply:
> >1. It more clearly indicates that literal and triple terms can not be used

> Cannot be used for what exactly?

As graph names.

> > 2. The intuition of variable substitution applies

> I am not sure what you mean by this comment.

In TriG, `GRAPH _:label { ... }` is legal so one can think of the query as a trig pattern, like BGPs are turtle patterns.

The variable may, visually, be defined in the query close-by.

(Yes, this actually happens when the "graph" evaluation is joined with the rest of the group graph pattern. Optimizers are prone to use known graph name though.)

Reply:
@afs 
>>> 1. It more clearly indicates that literal and triple terms can not be used
>>
>>    Cannot be used for what exactly?
>
> As graph names.
>
> It does not cause any harm to be a little bit more general here.
> The 1.1 text says "foreach IRI i in D"; the new text puts in a link so the cue isn't there.
> 
> "foreach [graph name](https://www.w3.org/TR/rdf12-concepts/#dfn-graph-name) gn in D"

Now I understand your concern! I had a proposal related to that already in my previous comment: We may extend this line of the algorithm as follows.

"foreach [graph name](https://www.w3.org/TR/rdf12-concepts/#dfn-graph-name) _gn_ in _D_ (recall that a graph name may be an IRI or a blank node)"

I think that should address your concern. Do you agree?

>>> 2. The intuition of variable substitution applies
>>
>> I am not sure what you mean by this comment.
>
> In TriG, `GRAPH _:label { ... }` is legal so one can think of the query as a trig pattern, like BGPs are turtle patterns.

Right.

Do you think my proposal for the previous point addresses this one as well? Otherwise, what else would you propose to do here?

Reply:
```suggestion
              an <a data-cite="RDF12-CONCEPTS#dfn-IRI">IRI</a>, or
```

Reply:
Just noting that `GRAPH _:blank` (if it were legal) is always empty. Blank nodes treatment as non-distinguished variables happens in BGPs.